### ISSUE TITLE
parent volumes are allowed to be removed from pv and pvc despite existing snapshots

### ISSUE THREAD

**Issue by zlangi** [ceph/ceph-csi#3593]
# Describe the bug #

The parent pvc can be deleted without issues with existing snapshots from the kubernetes cluster. It will disappear from the pvc and pv list, get moved to the trash on the ceph and the ceph will keep trying to remove it forever but fails to be removed due to existing snapshots.

# Environment details #

- Image/version of Ceph CSI driver : canary
- Helm chart version : canary
- Kernel version : 5.15.0-52-generic
- Mounter used for mounting PVC (for cephFS its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) : krbd
- Kubernetes cluster version : 1.23.12
- Ceph cluster version : 17.2.5

# Steps to reproduce #

1. Create a pvc on the kubernetes. Create snapshots from them, in my case it's 10 snapshots.
2. Delete the parent volume from the kubernetes with kubectl delete pvc which in my case will remove the pv as well.
3. The parent volume will disappear from the pvc and pv list also it will be moved to the trash where it will be attempted to be deleted
4. The ceph manager will report the following error:
`debug 2023-01-05T10:56:42.406+0000 7f59ddf24700  0 [rbd_support INFO root] execute_trash_remove: task={"sequence": 803528, "id": "71caadb1-a589-4c48-a162-153ce9f1d1ed", "message": "Removing image production-island11-overcloud/3127fdcbed096c from trash", "refs": {"action": "trash remove", "pool_name": "production-island11-overcloud", "pool_namespace": "", "image_id": "3127fdcbed096c"}, "retry_message": "[errno 39] RBD image has snapshots (error deleting image from trash)", "retry_attempts": 856, "retry_time": "2023-01-05T10:56:42.204701", "in_progress": true, "progress": 0.0}
`

# Expected behavior #

The expected behaviour is to not allow the parent to get deleted until all the snapshots were cleaned up and not to put it to the trash, because ceph will attempt to keep trying to delete and fails due to existing snapshots.

# Logs #

If the issue is in PVC creation, deletion, cloning please attach complete logs
of below containers.

- csi-provisioner and csi-rbdplugin/csi-cephfsplugin container logs from the
  provisioner pod.
`debug 2023-01-05T10:56:42.406+0000 7f59ddf24700  0 [rbd_support ERROR root] execute_task: [errno 39] RBD image has snapshots (error deleting image from trash)
debug 2023-01-05T10:56:42.406+0000 7f59ddf24700  0 [rbd_support INFO root] execute_task: task={"sequence": 803528, "id": "71caadb1-a589-4c48-a162-153ce9f1d1ed", "message": "Removing image production-island11-overcloud/3127fdcbed096c from trash", "refs": {"action": "trash remove", "pool_name": "production-island11-overcloud", "pool_namespace": "", "image_id": "3127fdcbed096c"}, "retry_message": "[errno 39] RBD image has snapshots (error deleting image from trash)", "retry_attempts": 856, "retry_time": "2023-01-05T10:56:42.204701"}
debug 2023-01-05T10:56:42.406+0000 7f59ddf24700  0 [rbd_support INFO root] execute_trash_remove: task={"sequence": 803528, "id": "71caadb1-a589-4c48-a162-153ce9f1d1ed", "message": "Removing image production-island11-overcloud/3127fdcbed096c from trash", "refs": {"action": "trash remove", "pool_name": "production-island11-overcloud", "pool_namespace": "", "image_id": "3127fdcbed096c"}, "retry_message": "[errno 39] RBD image has snapshots (error deleting image from trash)", "retry_attempts": 856, "retry_time": "2023-01-05T10:56:42.204701", "in_progress": true, "progress": 0.0}
`

Thanks in advance for looking into this!

**Comment by Rakshith-R** [comment]
@zlangi that is the expected behavior
Refer https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-snap-clone.md#delete-a-volume-pvc

**Comment by zlangi** [comment]
Ok thanks for the link, however we noticed a problem:
You delete all the pvc-s (in this case we talk about 20.000 PVC-s), each of them has 10 snapshots each.
Your pvc-s will disappear without issues, but your pv-s are not disappearing and get stuck in terminating state.
Let's have a look at a random pv-s output:

Name:            pvc-ffde0473-ffe6-4303-ac54-3a2ea3c4d804
Labels:          <none>
Annotations:     pv.kubernetes.io/provisioned-by: rbd.csi.ceph.com
                 volume.kubernetes.io/provisioner-deletion-secret-name: csi-rbd-secret
                 volume.kubernetes.io/provisioner-deletion-secret-namespace: infra
Finalizers:      [external-provisioner.volume.kubernetes.io/finalizer]
StorageClass:    ceph
Status:          Terminating (lasts 21m)
Claim:           default/stor-vol11481-claim
Reclaim Policy:  Delete
Access Modes:    RWO
VolumeMode:      Filesystem
Capacity:        50Gi
Node Affinity:   <none>
Message:         
Source:
    Type:              CSI (a Container Storage Interface (CSI) volume source)
    Driver:            rbd.csi.ceph.com
    FSType:            ext4
    VolumeHandle:      0001-0024-01cfdd82-965c-4d5f-bd5c-83b934bdd332-0000000000000003-8b6069af-13bf-4c68-ba30-079d5509e818
    ReadOnly:          false
    VolumeAttributes:      clusterID=01cfdd82-965c-4d5f-bd5c-83b934bdd332
                           imageFeatures=layering,exclusive-lock,object-map,fast-diff,deep-flatten
                           imageName=csi-vol-8b6069af-13bf-4c68-ba30-079d5509e818
                           journalPool=production-island11-overcloud
                           pool=production-island11-overcloud
                           storage.kubernetes.io/csiProvisionerIdentity=1671911636910-8081-rbd.csi.ceph.com
Events:
  Type     Reason              Age   From                                                                                            Message

  Warning  VolumeFailedDelete  58m   rbd.csi.ceph.com_ceph-csi-rbd-provisioner-ddbd49f65-qxcbm_cfac6fb2-5fca-462a-a809-cb7d1fd930e7  rpc error: code = Internal desc = rbd: ret=-34, Numerical result out of range
  Warning  VolumeFailedDelete  49m   rbd.csi.ceph.com_ceph-csi-rbd-provisioner-ddbd49f65-qxcbm_cfac6fb2-5fca-462a-a809-cb7d1fd930e7  rpc error: code = Internal desc = rbd: ret=-34, Numerical result out of range
  Warning  VolumeFailedDelete  19m   rbd.csi.ceph.com_ceph-csi-rbd-provisioner-ddbd49f65-qxcbm_cfac6fb2-5fca-462a-a809-cb7d1fd930e7  rpc error: code = Internal desc = rbd: ret=-34, Numerical result out of range

Basically the pv-s are stuck in terminating state despite the pvc-s are gone from the kubernetes cluster.
Is this a known issue?

**Comment by Madhu-1** [comment]
Same as https://github.com/ceph/go-ceph/issues/779 should be fixed once we update go-ceph to latest release https://github.com/ceph/go-ceph/releases/tag/v0.19.0

**Comment by zlangi** [comment]
Thanks, good to know that we hit this bug.
According to this commit: https://github.com/ceph/go-ceph/pull/785/files the limit was increased to 10240.
As far as I understand, even if I have the latest version with the limit of 10240 and if I try to delete my 20.000 pvc-s with snapshots on them, 9760 pv will be stuck in terminating state.
Also the pv-s which successfully made it to the trash won't be able to get deleted from trash until all their snapshots depending on them are removed so the ceph will keep trying to delete them again and again until finally all the snapshots cleaned up and the pv-s in the trash can be deleted and only after that the remaining 9760 pv-s will be moved to the trash and get out of their terminating state?

**Comment by vriabyk** [comment]
guys, sorry but I don't understand why you do allow this operation in ceph-csi. The user created volume in k8s, then created volumeSnapshot. Under the hood ceph rbd cloned the parent image. Then user deleted original volume from k8s, it passed successfully, but got stuck in ceph rbd trash. Is it expected behavior? I don't think so. I think ceph-csi shouldn't allow to delete original volume from k8s if it will stuck in ceph rbd trash. There is no 3rd party tools here, only k8s, ceph and ceph-csi between them.

**Comment by Rakshith-R** [comment]
> Thanks, good to know that we hit this bug. According to this commit: https://github.com/ceph/go-ceph/pull/785/files the limit was increased to 10240. As far as I understand, even if I have the latest version with the limit of 10240 and if I try to delete my 20.000 pvc-s with snapshots on them, 9760 pv will be stuck in terminating state. Also the pv-s which successfully made it to the trash won't be able to get deleted from trash until all their snapshots depending on them are removed so the ceph will keep trying to delete them again and again until finally all the snapshots cleaned up and the pv-s in the trash can be deleted and only after that the remaining 9760 pv-s will be moved to the trash and get out of their terminating state?

@zlangi 
I'll take care of this by adding a check to ignore out of range when we are checking trash. I'll send a fix soon, it should land in next release.


> guys, sorry but I don't understand why you do allow this operation in ceph-csi. The user created volume in k8s, then created volumeSnapshot. Under the hood ceph rbd cloned the parent image. Then user deleted original volume from k8s, it passed successfully, but got stuck in ceph rbd trash. Is it expected behavior? I don't think so. I think ceph-csi shouldn't allow to delete original volume from k8s if it will stuck in ceph rbd trash. There is no 3rd party tools here, only k8s, ceph and ceph-csi between them.

@vriabyk 
We use rbd layering feature, rbd child snap and clones are linked to the parent thereby avoiding a full copy(takes zero extra space).
It is not stuck in trash by mistake, it is by design, when all the linked child images are deleted, the parent in trash will automatically get deleted.
refer https://github.com/ceph/ceph-csi/blob/devel/docs/design/proposals/rbd-snap-clone.md#delete-a-volume-pvc  

This issue of `out of numerical range` which occurs with more than 1000 parent pvc in trash  will be taken care of soon.

**Comment by nixpanic** [comment]
This is probably blocked by ceph/ceph-csi#3593

**Comment by Rakshith-R** [comment]
> This is probably blocked by #3593

@nixpanic your comment refers back to this issue.
Did you mean to comment it on the pr or meant some other issue ?

**Comment by nixpanic** [comment]
On Mon, Jan 23, 2023 at 03:04:10AM -0800, Rakshith R wrote:
> > This is probably blocked by #3593
> 
> @nixpanic your comment refers back to this issue.
> Did you mean to comment it on the pr or meant some other issue ?

Hmm... I meant to link it to a go-ceph issue. But the only one related
seems to be closed. The PR increased the limit, but that still can be
reached.

**Comment by Rakshith-R** [comment]
As per discussion on pr https://github.com/ceph/ceph-csi/pull/3617

Just logging it as an error may not be the best approach.

We'll have to find another way to obtain the image id needed to add a task to remove it from trash other than going through all the images in the trash. 
I'll not be able to work on it anytime soon, unassigning so others can pick it up.
